lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -95,6 +95,179 @@ def self.persist_buffers!
+          AND (pea.object_attribute_id = d.child_id
+            OR pea.object_attribute_id = d.parent_id)
+          AND ra.id = pea.user_attribute_id
+          AND pea.operation_set_id = os.id
+          AND os.id = bb_a.parent_id
+          AND bb_a.child_id = crs_a.child_id
+          AND crs_a.parent_id = crs.id
+          AND crs_a.parent_id IN (#{crs_ids.join(", ")})
+        GROUP BY root_node, ra.operator_uri, crs.id
+        HAVING COUNT(DISTINCT bb_a.child_id) > 1
+      SQL
+
+      ancestors_query = <<-SQL
+        WITH RECURSIVE ancestors AS (
+          SELECT child_id, parent_id, child_id AS root_node
+          FROM assignments
+          WHERE child_id IN (#{operable_ids.join(", ")})
+          UNION ALL
+          SELECT a.child_id, a.parent_id, root_node
+          FROM assignments a
+            INNER JOIN ancestors ans
+            ON ans.parent_id = a.child_id
+          WHERE ans.parent_id = a.child_id
+        )
+        SELECT ans.root_node, ra.operator_uri, crs.id
+        FROM ancestors ans
+          JOIN policy_element_associations pea
+          ON pea.object_attribute_id = ans.child_id
+            OR pea.object_attribute_id = ans.parent_id
+          JOIN policy_elements as ra
+          ON ra.id = pea.user_attribute_id
+          JOIN policy_elements as os
+          ON pea.operation_set_id = os.id
+          JOIN assignments as bb_a
+          ON os.id = bb_a.parent_id
+          JOIN assignments as crs_a
+          ON bb_a.child_id = crs_a.child_id
+          JOIN policy_elements AS crs
+          ON crs_a.parent_id = crs.id
+        WHERE crs.active_policy_element_type = 'ConflictingRolesSet'
+          AND (pea.object_attribute_id = ans.child_id
+            OR pea.object_attribute_id = ans.parent_id)
+          AND ra.id = pea.user_attribute_id
+          AND pea.operation_set_id = os.id
+          AND os.id = bb_a.parent_id
+          AND bb_a.child_id = crs_a.child_id
+          AND crs_a.parent_id = crs.id
+          AND crs_a.parent_id IN (#{crs_ids.join(", ")})
+        GROUP BY root_node, ra.operator_uri, crs.id
+        HAVING COUNT(DISTINCT bb_a.child_id) > 1
+      SQL
+
+      result = PolicyElement.connection.exec_query(descendants_query).rows.flatten.map(&:to_i)
+      result.concat(PolicyElement.connection.exec_query(ancestors_query).rows.flatten.map(&:to_i))
+    end
+
+    # Check conflicts for Grant updates
+    def self.conflicts_for_operators_and_operables(operator_uri, operable_ids, crs_ids)
+      descendants_query = <<-SQL
+        WITH RECURSIVE descendants AS (
+          SELECT parent_id, child_id, parent_id AS root_node
+          FROM assignments
+          WHERE parent_id IN (#{operable_ids.join(", ")})
+          UNION ALL
+          SELECT a.parent_id, a.child_id, root_node
+          FROM assignments a
+            INNER JOIN descendants d
+            ON a.parent_id = d.child_id
+          WHERE a.parent_id = d.child_id
+        )
+        SELECT d.root_node, crs.id
```

Comment:
Strong preference to alias the `crs.id` result to `crs_id` or something more specific in the SELECT, so the output column isn't just `id`.